docs/CppInteroperability/ForwardVision.md

```diff
@@ -0,0 +1,290 @@
+
+## Appendix 1: Examples and Definitions
+
+**Reference Types** have reference semantics and object identity. A reference type is a pointer (or “reference”) to some object which means there is a layer of indirection. When a reference type is copied, the pointer’s value is copied rather than the object’s storage. This means reference types can be used to represent non-copyable types in C++. For real-world examples of C++ reference types, consider LLVM's [`Instruction` class](https://llvm.org/doxygen/IR_2Instruction_8h_source.html) or Qt's [`QWidget` class](https://github.com/qt/qtbase/blob/dev/src/widgets/kernel/qwidget.h).
+
+**Manually Managed Reference Types**
+
+Here a programmer has written a very large `StatefulObject` which contains many fields:
+
+```
+struct StatefulObject {
+  std::array<std::string, 32> names;
+  std::array<std::string, 32> places;
+  // ...
+  
+  StatefulObject(const StatefulObject&) = delete;
+  StatefulObject() = delete;
+
+  StatefulObject *create() { return new StatefulObject(); }
+};
+```
+
+
+Because this object is so expensive to copy, the programmer decided to delete the copy constructor. The programmer also decided that this object should be allocated on the heap, so they decided to delete the default constructor, and provide a create method in its place. 
+
+In Swift, this `StatefulObject` should be imported as a reference type, as it has reference semantics.
+
+**API Incorrectly Using Reference Types**
+
+Here someone has written an API that uses `StatefulObject` as a value type.
+
+```
+StatefulObject makeAppState();
+```
+
+This will invoke a copy of `StatefulObject` which violates the semantics that the API was written with. To be useable from Swift, this API needs to be updated to pass the object indirectly (by reference):
+
+```
+StatefulObject *makeAppState(); // OK
+const StatefulObject *makeAppState(); // OK
+StatefulObject &makeAppState(); // OK
+const StatefulObject &makeAppState(); // OK
+```
+
+**Immortal Reference Types**
+
+Instances of StatefulObject above are manually managed by the programmer, they create it with the create method and are responsible for destroying it once it is no longer needed. However, some reference types need to exist for the duration of the program, these references types are known as “immortal.” Examples of these immortal reference types might be pool allocators or app contexts. Let’s look at a GameContext object which allocates (and owns) various game elements:
+
+```
+struct GameContext {
+  // ...
+  
+  GameContext(const GameContext&) = delete;
+
+  Player *createPlayer();
+  Scene  *createScene();
+  Camera *createCamera();
+};
+```
+
+Here the GameContext is meant to last for the entire game as a global allocator/state. Because the context will never be deallocated, it is known as an “immortal reference type” and the Swift compiler can make certain assumptions about it. 
```

Comment:
formatting nit
```suggestion
Here the `GameContext` is meant to last for the entire game as a global allocator/state. Because the context will never be deallocated, it is known as an “immortal reference type” and the Swift compiler can make certain assumptions about it. 
```